### PR TITLE
Bug 1786314: Bump dependencies

### PR DESCRIPTION
**Pull request by pierreprinetti**
* github.com/gophercloud/gophercloud
* github.com/gophercloud/utils
* github.com/terraform-provider-openstack/terraform-provider-openstack

and dependant packages.

/label platform/openstack

### CODE REVIEW

**Comment by pierreprinetti** [comment]
Needs https://github.com/openshift/hashicorp-terraform-plugin-sdk/pull/1

**Review by Fedosin — COMMENTED** [review]
/test e2e-gcp
/test e2e-azure

File: pkg/destroy/openstack/openstack.go
Comment (mandre) [inline pkg/destroy/openstack/openstack.go:262]: I would expect the commit message to explain why this change was needed, even perhaps splitting this into a separate commit.

Comment (Fedosin) [reply pkg/destroy/openstack/openstack.go:262]: I believe the reason is https://github.com/gophercloud/gophercloud/pull/2070

Comment (pierreprinetti) [reply pkg/destroy/openstack/openstack.go:262]: Thanks Mike, that's correct. I have updated the commit message with that information.

**Review by mandre — COMMENTED** [review]
/lgtm

**Comment by pierreprinetti** [comment]
/assign @crawford

just as the bot suggests.

**Review by Fedosin — APPROVED** [review]
/lgtm

**Comment by Fedosin** [comment]
/hold

bumping golang to 1.15 breaks the installer: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/15083/rehearse-15083-pull-ci-openshift-installer-master-unit/1351880899611856896

**Comment by pierreprinetti** [comment]
Removed the 1.15 bump. Thanks @Fedosin 
/hold cancel

**Comment by pierreprinetti** [comment]
/cc @staebler 
FYI, since you approved the previous bump.

**Comment by staebler** [comment]
I am not comfortable making this many changes to the installer's dependencies this late in the 4.7 release cycle, particularly to fix a bug that is low severity.

**Comment by pierreprinetti** [comment]
> I am not comfortable making this many changes to the installer's dependencies this late in the 4.7 release cycle, particularly to fix a bug that is low severity.

👌

/hold until master is open again

**Comment by staebler** [comment]
There are some conflicts that need to be resolved now.

Out of curiosity, what were the commands that you ran to bump the gophercloud dependency? It seems to have a big ripple through the rest of the dependencies that I find somewhat surprising.

**Comment by pierreprinetti** [comment]
@staebler 
It all starts with this, from the current master:

```
$ go get -u github.com/gophercloud/gophercloud github.com/gophercloud/utils
go: github.com/gophercloud/utils upgrade => v0.0.0-20210209042946-13abf2251886
go: github.com/gophercloud/gophercloud upgrade => v0.15.0

$ go mod vendor
warning: ignoring symlink /home/pierre/code/src/github.com/openshift/installer/pkg/asset/store/data
go: finding module for package github.com/gophercloud/utils/openstack/compute/v2/images
github.com/openshift/installer/pkg/terraform/exec/plugins imports
        github.com/terraform-provider-openstack/terraform-provider-openstack/openstack imports
        github.com/gophercloud/utils/openstack/compute/v2/images: module github.com/gophercloud/utils@latest found (v0.0.0-20210209042946-13abf2251886), but does not contain package github.com/gophercloud/utils/openstack/compute/v2/images
```

Bumping gophercloud breaks the Terraform provider, which must then be updated.

```
$ go get -u github.com/terraform-provider-openstack/terraform-provider-openstack

[...]
# cloud.google.com/go/storage
../../../../pkg/mod/cloud.google.com/go/storage@v1.13.0/storage.go:124:23: undefined: internaloption.WithDefaultMTLSEndpoint
../../../../pkg/mod/cloud.google.com/go/storage@v1.13.0/storage.go:132:23: undefined: internaloption.WithDefaultMTLSEndpoint
# github.com/hashicorp/terraform-plugin-sdk/internal/configs/configload
../../../../pkg/mod/github.com/openshift/hashicorp-terraform-plugin-sdk@v1.14.0-openshift/internal/configs/configload/loader_snapshot.go:52:29: cannot use fs (type snapshotFS) as type afero.Fs in argument to configs.NewParser:
        snapshotFS does not implement afero.Fs (missing Chown method)
../../../../pkg/mod/github.com/openshift/hashicorp-terraform-plugin-sdk@v1.14.0-openshift/internal/configs/configload/loader_snapshot.go:57:28: cannot use fs (type snapshotFS) as type afero.Fs in field value:
        snapshotFS does not implement afero.Fs (missing Chown method)
../../../../pkg/mod/github.com/openshift/hashicorp-terraform-plugin-sdk@v1.14.0-openshift/internal/configs/configload/loader_snapshot.go:200:5: cannot use snapshotFS literal (type snapshotFS) as type afero.Fs in assignment:
        snapshotFS does not implement afero.Fs (missing Chown method)
```

The Terraform bump triggers the upgrade of `github.com/spf13/afero`, which breaks `github.com/openshift/hashicorp-terraform-plugin-sdk`.

The chain continues in a spiral of pain and despair.

---

One problem I see here is that several of our depedencies do not respect the "breaking change -> major bump" rule of semver, on which gomod's minimal version selection is based. I was very surprised to see Steve Francia's `afero` changing an exported interface without major bump. My personal learning is that it is important to influence library maintainers to drop the fear of bumping major on "small" breaking changes.

---

Would you have done differently?

**Comment by staebler** [comment]
> Would you have done differently?

@pierreprinetti The larger scope of the changes is due to using the `-u` flag in the `go get`. That flag tells go to update all of the dependencies of the module to the latest as well.

I was able to build successfully with the following changes to go.mod and an update of a call in the destroyer.
```
$ git diff go.mod
diff --git a/go.mod b/go.mod
index 0543098ac..80f7a3393 100644
--- a/go.mod
+++ b/go.mod
@@ -32,8 +32,8 @@ require (
        github.com/golang/protobuf v1.4.3
        github.com/google/martian v2.1.1-0.20190517191504-25dcb96d9e51+incompatible // indirect
        github.com/google/uuid v1.1.2
-       github.com/gophercloud/gophercloud v0.12.1-0.20200827191144-bb4781e9de45
-       github.com/gophercloud/utils v0.0.0-20201212031956-9dc30e126fea
+       github.com/gophercloud/gophercloud v0.15.1-0.20210202035223-633d73521055
+       github.com/gophercloud/utils v0.0.0-20210209042946-13abf2251886
        github.com/h2non/filetype v1.0.12
        github.com/hashicorp/go-azure-helpers v0.10.0
        github.com/hashicorp/go-plugin v1.3.0
@@ -77,7 +77,7 @@ require (
        github.com/sirupsen/logrus v1.6.0
        github.com/spf13/cobra v1.1.1
        github.com/stretchr/testify v1.6.1
-       github.com/terraform-provider-openstack/terraform-provider-openstack v1.33.0
+       github.com/terraform-provider-openstack/terraform-provider-openstack v1.37.0
        github.com/terraform-providers/terraform-provider-aws v1.60.1-0.20200807230610-d5346d47e3af
        github.com/terraform-providers/terraform-provider-azurerm v1.44.1-0.20200911233120-57b2bfc9d42c
        github.com/terraform-providers/terraform-provider-google v1.20.1-0.20200623174414-27107f2ee160
```

**Comment by pierreprinetti** [comment]
@staebler 
So instead of using ‘go get -u’, you changed the versions by hand in ‘go.mod’?

Honestly I did see some value in upgrading more packages. If you prefer the upgrade to be restricted to as few packages as possible, then I’ll apply your diff instead.

**Comment by staebler** [comment]
> @staebler
> So instead of using ‘go get -u’, you changed the versions by hand in ‘go.mod’?
> 
> Honestly I did see some value in upgrading more packages. If you prefer the upgrade to be restricted to as few packages as possible, then I’ll apply your diff instead.

Personally, yes, I do update the go.mod by hand. But you do not have to. By point was not that you should not use `go get`. My point was that you will get fewer upgraded modules if you omit the `-u` flag

```
$ go get github.com/gophercloud/gophercloud
go: github.com/gophercloud/gophercloud upgrade => v0.15.0
go: downloading github.com/gophercloud/gophercloud v0.15.0

$ go get github.com/terraform-provider-openstack/terraform-provider-openstack
go: github.com/terraform-provider-openstack/terraform-provider-openstack upgrade => v1.37.0
go: downloading github.com/gophercloud/utils v0.0.0-20210202040619-eca783186fc4
```

And, yes, please keep the changes limited to what you need.

**Comment by pierreprinetti** [comment]
@staebler PTAL

**Comment by staebler** [comment]
/hold

This looks like a legitimate failure caused by these changes.

```
level=fatal msg=failed to fetch Master Machines: failed to load asset "Install Config": failed to create a network client: unable to load secure.yaml: no secure.yml file found
```

**Comment by EmilienM** [comment]
I think we hit https://github.com/gophercloud/utils/blob/f6de111f2eaedc778f842096b07cbc319a7b3cdd/openstack/clientconfig/utils.go#L169 so maybe we need to create that file in our CI? Although it sounds backward incompatible change for our users who might not have secure.yaml.

**Review by mandre — COMMENTED** [review]
/lgtm

**Comment by EmilienM** [comment]
@staebler finally passing CI, any change to approve it please?

Thanks

**Comment by staebler** [comment]
> @staebler finally passing CI, any change to approve it please?
> 
> Thanks

This has already been approved. I didn't mean to delay this PR with the hold: I only wanted to keep the tests from retrying.

/hold cancel

**Pull request merged**
